splink/internals/linker_components/clustering.py

```diff
@@ -334,6 +346,307 @@ def cluster_using_single_best_links(
+        `linker.inference.predict()` into groups of connected record using the connected
+        components graph clustering algorithm
+
+        Records with an estimated `match_probability` at or above each of the values in
+        `threshold_match_probabilities` (or records with a `match_weight` at or above
+        each of the values in `threshold_match_weights`) are considered to be a match
+        (i.e. they represent the same entity).
+
+        This function efficiently computes clusters for multiple thresholds by starting
+        with the lowest threshold and incrementally updating clusters for higher
+         thresholds.
+
+        Args:
+            df_predict (SplinkDataFrame): The results of `linker.predict()`
+            threshold_match_probabilities (list[float] | None): List of match
+                probability thresholds to compute clusters for
+            threshold_match_weights (list[float] | None): List of match weight
+                 thresholds to compute clusters for
+
+        Returns:
+            SplinkDataFrame: A SplinkDataFrame containing a list of all IDs, clustered
+                into groups for each of the desired match thresholds.
+
+                The output dataframe will contain the following metadata:
+
+                - threshold_match_probabilities: List of match probability thresholds
+
+                - cluster_summary_stats: summary statistics (number of clusters, max
+                 cluster size, avg cluster size) for each threshold
+
+        Examples:
+            ```python
+            df_predict = linker.inference.predict(threshold_match_probability=0.5)
+            df_clustered =
+                linker.clustering.cluster_pairwise_predictions_at_multiple_thresholds(
+                df_predict, threshold_match_probability=0.95
+            )
+
+            # Access the match probability thresholds
+            match_prob_thresholds = df_clustered
+                .metadata["threshold_match_probabilities"]
+
+            # Access the cluster summary stats
+            cluster_summary_stats = df_clustered.metadata["cluster_summary_stats"]
+            ```
+
+        """
+
+        # Strategy to cluster at multiple thresholds:
+        # 1. Cluster at the lowest threshold
+        # 2. At next threshold, note that some clusters do not need to be recomputed.
+        #    Specifically, those where the min probability within the cluster is
+        #    greater than this next threshold.
+        # 3. Compute remaining clusters which _are_ affected by the next threshold.
+        # 4. Repeat for remaining thresholds.
+
+        # Need to get nodes and edges in a format suitable to pass to
+        # cluster_pairwise_predictions_at_threshold
+        linker = self._linker
+        db_api = linker._db_api
+
+        pipeline = CTEPipeline()
+
+        enqueue_df_concat(linker, pipeline)
+
+        uid_cols = linker._settings_obj.column_info_settings.unique_id_input_columns
```

Comment:
I'm wondering whether we should factor out the 'setup' here into a function that's shared between `linker.cluster_pairwise_predictions_at_threshold` and `linker.cluster_pairwise_predictions_at_multiple_thresholds`

Or - should cluster_pairwise_predictions_at_threshold simply call `linker.cluster_pairwise_predictions_at_multiple_thresholds` with a single threhold value

Reply:
Yea I did think that moving to the multiple threshold as the default engine could be helpful in reducing duplication. I will implement when I am back from holiday